bin/run_clang_tidy_check

```diff
@@ -0,0 +1,128 @@
+# compile_commands.json files should be present somewhere in the tree (they
+# don't have to be committed to git). These can be generated by CMake by passing
+# the argument -DCMAKE_EXPORT_COMPILE_COMMANDS=1.
+#
+# If the argument --generate-fixes is passed to this script, then it will
+# deposit a series of YAML files into ./clang_tidy_fixes. These can then be
+# applied like so:
+#       clang-apply-replacements ./clang_tidy_fixes
+#
+# Note that only a small subset of the checks are actually capable of generating
+# fixes at present (though most of the modernize-* checks seem to work).
+
+import collections
+import itertools
+import json
+import logging
+import multiprocessing as mp
+import os
+import re
+import subprocess
+import sys
+from xml.sax.saxutils import escape
+
+INCLUDE_CHECKS = (
+    'bugprone-*',
+    'clang-analyzer-*',
+    'concurrency-*',
+    'misc-*',
+    'modernize-*',
+    'openmp-*',
+    'performance-*',
+)
+
+# Some of these are a bit overzealous...
+EXCLUDE_CHECKS = (
+    '*osx.*',
+    'bugprone-branch-clone',
+    'bugprone-exception-escape',
+    'bugprone-integer-division',
+    'bugprone-narrowing-conversions',
+    'clang-analyzer-webkit*',
+    'misc-non-private-member-variables-in-classes',
+    'modernize-avoid-c-arrays',
+    'modernize-concat-nested-namespaces',
+    'modernize-pass-by-value',
+    'modernize-use-auto',
+    'modernize-use-emplace',
+    'modernize-use-equals-default',
+    'modernize-use-nodiscard',
+    'modernize-use-override',
+    'modernize-raw-string-literal',
+    'modernize-use-trailing-return-type',
+    'modernize-return-braced-init-list',
+    'openmp-use-default-none',
+    'performance-type-promotion-in-math-fn',
+)
+check_args = None
+if INCLUDE_CHECKS or EXCLUDE_CHECKS:
+    checks = itertools.chain(INCLUDE_CHECKS, ('-' + check for check in EXCLUDE_CHECKS))
+    check_args = '-checks=' + ','.join(checks)
+
+# List the currently enabled checks
+args = ['/usr/bin/clang-tidy', '-list-checks']
+if check_args:
+    args.append(check_args)
+subprocess.run(args)
+
+generate_fixes = len(sys.argv) > 1 and sys.argv[1] == '--generate-fixes'
+FIXES_DIR = 'clang_tidy_fixes'
+if generate_fixes and not os.path.isdir('fixes'):
+    os.mkdir(FIXES_DIR)
+
+def run_clang_tidy(file):
+    args = ['/usr/bin/clang-tidy', '-p', '.', file]
+
+    # Explicitly (disallow) various checks
+    if check_args:
+        args.append(check_args)
+
+    # Optionally generate fixes
+    if generate_fixes:
+        fixes_file = os.path.join(FIXES_DIR, file.replace('/', '_') + '.yaml')
+        args.append('--export-fixes=' + fixes_file)
+
+    proc = subprocess.Popen(args, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+    stdout = proc.communicate()[0].decode('utf-8')
+    if stdout != '':
+        print(stdout)
+
+if __name__ == '__main__':
+    all_json = []
+    all_files = set()
+    exclude_path = os.path.join(os.getcwd(), 'third_party')
+    for root, _, files in os.walk('.'):
```

Comment:
I think you could get rid of quite a lot of this logic with ``glob.glob("**/compile_commands.json", recursive=True``

Reply:
So you can! I should have guessed there'd be a more pythonic way to do that...

Reply:
Hah, for any given piece of code and a personal view on pythonicness, there _always_ is....